dhall/src/Dhall/Repl.hs

```diff
@@ -234,6 +235,10 @@ saveBinding (file : "=" : tokens) = do
   liftIO (System.IO.withFile file System.IO.WriteMode handler)
 saveBinding _ = fail ":save should be of the form `:save x = y`"
 
+cmdQuit :: ( MonadIO m, MonadState Env m ) => [String] -> m ()
+cmdQuit _ = do
+  liftIO (putStrLn "Goodbye.")
+  throw Interrupt
```

Comment:
You usually want to prefer `throwIO` instead of `throw` since `throw` can accidentally be triggered by evaluation whereas `throwIO` is evaluation-safe.  Or in other words:

```haskell
throw x `seq` () = throw x

throwIO x `seq` () = ()
```

```suggestion
  liftIO (throwIO Interrupt)
```

Reply:
ah so technically it is possible it could throw without `putStrLn`ing? Updated.

Reply:
Yeah!  The `repline` loop has logic that checks if the last command was non-empty when an `Interrupt` was thrown.  If the command is empty then it says `Goodbye.` and if it was non-empty then it says nothing.  This is why you had to insert an explicit `Goodbye.` to match the behavior because `:quit` was a non-empty command.